pymatgen/io/vasp/inputs.py

```diff
@@ -964,29 +986,44 @@ def __add__(self, other):
 
     def check_params(self):
         """
-        Raises a warning for nonsensical or non-existent INCAR tags and
-        parameters. If a keyword doesn't exist (e.g. there's a typo in a
-        keyword), your calculation will still run, however VASP will ignore the
-        parameter without letting you know, hence why we have this Incar method.
+        Check INCAR for invalid tags or values.
+        If a tag doesn't exist, calculation will still run, however VASP
+        will ignore the tag and set it as default without letting you know.
         """
-        for key, val in self.items():
-            # First check if this parameter even exists
-            if key not in incar_params:
-                warnings.warn(f"Cannot find {key} in the list of INCAR flags", BadIncarWarning, stacklevel=2)
+        # Load INCAR tag/value check reference file
+        with open(f"{module_dir}/incar_parameters.json", encoding="utf-8") \
+                as json_file:
+            incar_params = json.loads(json_file.read())
+
+        for tag, val in self.items():
+            # Check if the tag exists
+            if tag not in incar_params:
+                warnings.warn(
+                    f"Cannot find {tag} in the list of INCAR tags",
+                    BadIncarWarning, stacklevel=2
+                    )
+                continue
 
-            if key in incar_params:
-                if type(incar_params[key]).__name__ == "str":
-                    # Now we check if this is an appropriate parameter type
-                    if incar_params[key] == "float":
-                        if not type(val) not in ["float", "int"]:
-                            warnings.warn(f"{key}: {val} is not real", BadIncarWarning, stacklevel=2)
-                    elif type(val).__name__ != incar_params[key]:
-                        warnings.warn(f"{key}: {val} is not a {incar_params[key]}", BadIncarWarning, stacklevel=2)
+            # Check value and its type
+            param_type = incar_params[tag].get("type")
+            allowed_values = incar_params[tag].get("values")
 
-                # if we have a list of possible parameters, check
-                # if the user given parameter is in this list
-                elif type(incar_params[key]).__name__ == "list" and val not in incar_params[key]:
-                    warnings.warn(f"{key}: Cannot find {val} in the list of parameters", BadIncarWarning, stacklevel=2)
+            if param_type is not None \
+                    and not isinstance(val, eval(param_type)):
```

Comment:
`eval()` might potentially bring security vulnerabilities but as we're not reading input from users here, it should be safe. For some reason `literal_eval` doesn't work and I would see if I could fix it.